joomla/joomla-cms
# [5.2] mod_redirect - Admin Module to display Redirects: Links

### Summary of Changes
This is PR creates **mod_redirect**, a new **Admin Module** that displays "**Redirects: Links**", the links that are displayed in the "Redirects" component. 

**Example output**:
![dashboard](https://github.com/user-attachments/assets/cb42f8a2-05a3-46c2-9449-95072b3129dc)


### Background
Broken links have a negative impact on the user experience and also harm SEO. 
Joomla has in the core an excellent Redirects component that administrators can use to manually create "301 redirects" for faulty links. 

Administrators can manage faulty links via: System > (under Manage) Redirects. 
It makes sense to redirect the most visited unredirected faulty links to working pages.
To display these links in the Redirects component:
- Go to [Filter Options]
- Set State to "Disabled"
- Sort the table by "404 Hits descending"

![redirects-links-component](https://github.com/user-attachments/assets/9653068c-37a2-4f45-a5db-c974f1bff81e)

### Admin Module: mod_redirect
To simplify the process for admins, this new Redirect Admin Module displays a quick overview of faulty links that can be added to a Dashboard. 

By default, the module displays:
- The top 5 redirects links
- Links with a "disabled" state
- Links with an HTTP status code of "301"
- The list is sorted by the number of hits in descending order, with the most frequently hit links at the top.

This way it's easier to see which faulty links in the website need to be redirected.  

### Installation
Add the PR to your local environment (e.g. using the Joomla Patchtester Component)

**Install the module** via: System > **Discover** > install "Redirects: Links".
- When this module is added to the Joomla core, we need to update the #__extensions table via /installation/sql/mysql/base.sql + /installation/sql/postgresql/base.sql.
- This module should NOT be added to the dashboard by default because the Redirect Plugin is not enabled by default.

![extension-discover](https://github.com/user-attachments/assets/5b01551d-51d5-4ae6-841f-8b82e406ebf7)

**Create a New module** via: Content > Administrator Modules, New 

![module-new](https://github.com/user-attachments/assets/e7b79935-ce5d-49ea-b06a-9e3853bc3691)

**Configure the Module**, add a Title, set the Position to "cpanel" and status to "Published,"
 
![module-config](https://github.com/user-attachments/assets/b7aa5148-f6b5-44e7-a865-f470f5af6bc8)

### Sample data
**To test** the configuration and output of **the module**, the **Redirects Component needs to have some links**. 

You can create your own faulty links within Joomla:
- The Redirect Plugin must be enabled
- In the Redirect component, the 'Collect URLs' option should be enabled
- Visit faulty links on the front-end to generate entries in the Redirects component

Or you can use the following **SQL sample data** as "Redirects: Links".
Run the following SQL in your database (e.g. using phpMyAdmin) to create sample records.
You'll need to change the #__ prefix with your Joomla's database prefix.

```sql
INSERT INTO `#__redirect_links` (`old_url`, `new_url`, `referer`, `comment`, `hits`, `published`, `created_date`, `modified_date`, `header`) VALUES
('https://local.test.nl/faulty-link', NULL, '', '', 101, 0, '2024-08-08 08:13:29', '2024-08-11 08:10:02', 301),
('https://local.test.nl/faulty-link-archived', NULL, '', '', 1, 2, '2024-08-05 08:10:08', '2024-08-11 08:10:08', 301),
('https://local.test.nl/faulty-link-trashed', NULL, '', '', 1, -2, '2023-08-06 08:10:15', '2024-08-11 08:10:15', 301),
('https://local.test.nl/faulty-link-with-xss-attempt&param=<script>alert(\'xss\')</script>', NULL, '', '', 1001, 0, '2023-07-23 12:11:26', '2024-08-11 08:11:26', 301),
('https://local.test.nl/faulty-link-disabled', NULL, '', '', 1, 0, '2021-01-21 18:12:45', '2024-08-11 08:12:45', 301),
('https://local.test.nl/faulty-link-redirected', 'https://local.test.nl/', '', 'redirect to home', 13, 1, '2024-08-08 08:13:29', '2024-08-11 08:13:29', 301),
('https://local.test.nl/another-faulty-link', NULL, '', '', 1, 0, '2024-08-11 08:13:47', '2024-08-11 08:13:47', 301),
('https://local.test.nl/yet-another-faulty-link', NULL, '', '', 1, 0, '2023-01-01 01:02:03', '2024-08-11 08:13:52', 301),
('https://local.test.nl/what-is-wrong-at-this-site', NULL, '', '', 11, 0, '2024-08-11 08:14:07', '2024-08-11 08:14:07', 301),
('https://local.test.nl/pe7er-was-here', NULL, '', '', 93258, 0, '2024-08-11 10:10:10', '2024-08-11 08:14:15', 301);
```

### Testing Instructions
- Test changes in the Redirect component. Modify the state of some links. Check how these changes reflect in the "Redirects: Links" Module.
- Test changes in the module configuration. Check if those are displayed correctly on the Dashboard
- Test that the displayed links are sanitized against vulnerabilities like JavaScript URL injection.

### Link to documentations
Please select:
- [x] Documentation link for docs.joomla.org: **Documentation needs to be created. I'll create documentation after this feature has been merged**
